Created API endpoint for GET request (list view) for USERS with proper JWT authentication

Esterello2 opened the pull request:
Modified views.py, urls.py, and serialization.py to establish endpoint for GET User (List View) with JWT authentication. This included adding the simpleJWT module from rest framework, along with JWT token built-in API views to allow the for the generation and refresh of tokens.
A POST User request API endpoint was also created here, which was also originally meant to deal with JWT authentication, however, team decided to make it an open endpoint.

Esterello2 committed:
Create serializers.py

Specifies how user data should be serialized

Esterello2 committed:
Update urls.py

added UserCreateView to allow for the POST of users

Esterello2 committed:
Update views.py

added a view to deal with creating a new User

Esterello2 committed:
Update views.py

Added a check for authenticated users only to be able to send GET, POST, and PATCH requests for User info.

Esterello2 committed:
Update

Esterello2 committed:
Deleted some files

Esterello2 commented:
Updated the PR by deleting the old files from the "users" directory, and adding the serializers, views, urls, and permissions files in a directory called API.

Esterello2 committed:
Merge branch 'dev' of https://github.com/uchicago-cs/chigame into apis/post-users

Esterello2 commented:
Made a big update here, I added simple JWT authentication by importing the simpleJWT module from the rest_framework app. I incorporated three new views that deal with generating, refreshing, and verifying JWT tokens to properly authenticate API requests for POST and GET for User info.

cuyakonwu commented on src/chigame/api/permissions.py:
Any reason you check for unauthentication as opposed to authentication?

Esterello2 replied:
Thanks for pointing this out, it was originally done like this because the only POST requests for Users I had in mind were when new users were creating accounts. I will remove this and update with the new implementation we agreed on that only admin users can create new users through API request.

cuyakonwu commented on src/chigame/api/serializers.py:
Glad you condensed the user serializers! Any reason for the notification serializer in this PR specifically?

Esterello2 replied:
Thank you for pointing this out, I was actually testing this out by making a serializer for the Notification model, but I will actually look to remove this now to keep the PR organized.

cuyakonwu commented on src/chigame/api/urls.py:
Please leave the comments in, it helps with organization.

cuyakonwu reviewed (approved):
Looks good generally! Please try to get back to the comments I made on the code though.

Esterello2 committed:
Fixed CI error

Esterello2 committed:
Merge branch 'dev' into apis/post-users

Esterello2 commented:
Added proper JWT authentication to a POST User API endpoint that allows a successful POST only if a valid JWT token is provided. Also, incorporated in the User List View all to ensure only Authenticated entities can see all users info. Deleted the extraneous code Conrad mentioned.

Esterello2 committed:
Fixing CI issues

Esterello2 committed:
fixing CI issue in urls.py

Esterello2 committed:
Fixing CI issues

Esterello2 committed:
Fixing CI

Esterello2 committed:
Fix CI

Esterello2 committed:
CI ISsue

Esterello2 committed:
Fixing CI

Esterello2 committed:
sd

Esterello2 committed:
djsdf

Esterello2 committed:
update:

Esterello2 committed:
Merge branch 'dev' into apis/post-users

Esterello2 committed:
update

majorsylvie committed:
Merge branch 'dev' into apis/post-users

Esterello2 committed:
Merge branch 'dev' into apis/post-users

Esterello2 committed:
CI update

Esterello2 committed:
Merge branch 'dev' of https://github.com/uchicago-cs/chigame into apis/post-users

Esterello2 committed:
update

Esterello2 commented:
Made a final edit after consulting team to take off JWT Authentication method for POST new users, and only kept it for ListView of Users.

Esterello2 committed:
update

Esterello2 committed:
update

Esterello2 committed:
ipdate

Esterello2 committed:
Merge branch 'dev' into apis/post-users

Esterello2 committed:
linter

Esterello2 committed:
updated

Esterello2 committed:
update

Esterello2 committed:
linting

majorsylvie committed:
Merge branch 'dev' into apis/post-users

majorsylvie committed:
Merge branch 'dev' into apis/post-users

Esterello2 commented:
A quick overview of using Postman to test JWT authentication on the GET User listview API endpoint:

1. To generate a token, you have to go to the api/token/ url endpoint in the DRF GUI endpoint.
2. Enter superuser credentials (email: [esteban@test.com](mailto:esteban@test.com) pass: Estetheyan)
3. It will return two token strings, one that says refresh and other that says accept
4. Copy the "accept" one, and then this is the part that so far can only be done on Postman interface
5. Postman lets you send a GET request to a protected endpoint with the token you copied.
6. Enter the appropriate URL (~/api/users/)
7. In the Auth tab in the Postman interface, select the dropdown under type that says "Bearer Token", and paste token string into the "Token" field on the right.
8. Ensuring "GET" is selected as Request type, select Send.

https://github.com/uchicago-cs/chigame/assets/100042294/b1b5ab9e-c717-42dd-9f36-c3ff511a88ab

Esterello2 committed:
Merge branch 'dev' into apis/post-users

majorsylvie commented on src/chigame/api/urls.py:
```diff
@@ -3,17 +3,19 @@
+    path("token/verify/", views.CustomTokenVerifyView.as_view(), name="token_verify"),
     path("users/<slug:slug>/", views.UserDetailView.as_view(), name="api-user-detail"),
     path("users/<int:pk>/friends/", views.UserFriendsAPIView.as_view(), name="api-user-friends"),
-    # CHAT API URLS
```
Why remove these comments? they are relevant and good for organization.

majorsylvie commented on src/chigame/api/urls.py:
```diff
@@ -3,17 +3,19 @@
-    # LOBBY API URLS
+    path("tournaments/create/", views.TournamentCreateView.as_view(), name="create-tournament"),
+    path("tournaments/", views.TournamentListView.as_view(), name="tournament-list"),
+    path("register/", views.UserRegistrationView.as_view(), name="user-registration"),
```
What is the point of this register endpoint?
Upon trying to access it it, it doesn't allow get. 
You don't mention it in your JWT walkthrough, nor have any code comments explaining this.
Please either remove this URL, or explain and justify why this endpoint should be in this PR!

Esterello2 replied:
The register endpoint is the POST new user endpoint that does not have JWT, should I name it something more clear?

majorsylvie replied:
Yes please!

majorsylvie replied:
could also be `api/users`register`

majorsylvie commented on src/chigame/api/views.py:
```diff
@@ -42,6 +49,39 @@ class GameDetailView(generics.RetrieveUpdateDestroyAPIView):
+    permission_classes = [IsAdminUser]
+
+
+class TournamentListView(generics.ListAPIView):
```
Please remove this tournament code

Esterello2 replied:
I deleted this

majorsylvie reviewed (changes requested):
I can successfully NOT access `/api/users` when I don't use JWT authentication.

Yet I can still POST users (as you've said the team has decided to allow) 

Then I can successfully view the `api/users` when GET requesting the endpoint with a proper JWT token, as generated from `api/token` (and verifiable as a successful token by visiting `api/token/verify` and POSTing the `access` token I got from `api/token`.

There are multiple comments about extraneous code. Please resolve then we can merge your actual functionality!

majorsylvie committed:
Merge branch 'dev' into apis/post-users

Esterello2 committed:
Merge branch 'dev' into apis/post-users

Esterello2 committed:
update

majorsylvie reviewed (changes requested):
Seeing that you removed the tournament code from this PR, this is almost good to merge! Please add a code comment and make the `register` endpoint named more clearly. 

Other than that, given the JWT GET users functionality is working as it's supposed to, this will be good to merge!

cuyakonwu committed:
Merge branch 'dev' into apis/post-users

majorsylvie committed:
Merge branch 'dev' into apis/post-users
